tensorflow/python/keras/layers/convolutional.py

```diff
@@ -94,6 +94,7 @@ class Conv(Layer):
     trainable: Boolean, if `True` the weights of this layer will be marked as
       trainable (and listed in `layer.trainable_weights`).
     name: A string, the name of the layer.
+    fused: Use fused dilated convolution kernel. This is only supported on GPUs.
```

Comment:
It would be better to select the optimal behavior, i.e. fused when possible, automatically, and then there would be no need for this additional option.

Is there any reason that can't be done?

Reply:
CPU doesn't support directed dilated convolution. But GPU supports that. The reason why using `fused` is because we cannot get the availability of the GPUs inside the OP (although in eager mode, we can use context.num_gpus()), which is the critical information to decide whether we can call the direct dilated convolution or not.

Reply:
That's a good point.  In practice the user will be constructing the graph for a particular device, but I suppose that information may not be available to tensorflow itself, since the user may be using a default device placement.

I do think it would be pretty unfortunate to have to default this optimization to off and require the user to manually set it, but it seems like that may be a bit tricky to achieve.

I suppose one option would be to incorporate the space-to-batch transform inside the CPU convolution implementation itself, but that would be a fair amount of work and a somewhat unfortunate workaround.

I think we could benefit from someone with more knowledge of Tensorflow than me to weigh in on this regarding the best way to achieve the automatic selection behavior.

Reply:
Yes, you are right. In 2.0, we do have the so called implementation selector (which is a grappler pass before function optimizer). I am actually working on it, but I got some problem of `mismatched signatures` between the CPU and GPU methods. @reedwm mentioned someone from Google could help with that. So, I might need some help from you guys before my next commits.

Reply:
Try doing one of the following:
1. Only use the implementation selector if mixed precision and distribution strategy are not being used. This is done in GRU V2 [here](https://github.com/tensorflow/tensorflow/blob/845b6f722147184b6841f58f96d670b094ddee61/tensorflow/python/keras/layers/recurrent_v2.py#L374-L382)
2. When you create the two Functions for the implementation_selector, do not pass in any variables. Instead, pass in the result of Variable.read_value(). Also do not capture any variables or tensors. implementation_selector can be buggy when DistributedVariables or AutoCastVariables are used, or when tensors are captured.
3. Support dilated convolution on CPUs (only if you have time)
4. Instead of using FusedBatchNorm, create a fused layer norm op. But I think you'll need to implement it for CPUs, GPUs, XLA, and perhaps other backends like TFLite and TensorRT, so don't attempt this unless you have time.

If neither 1 nor 2 work, let me know.

Reply:
I tried the method 1 in a separate branch https://github.com/tensorflow/tensorflow/compare/master...houtoms:test-impl-selector?expand=1. The eager version works find but the graph one doesn't work. Basically, in the nn_ops.py you can see 
```        
        conv_op_standard = _generate_defun_backend(
            api_name, _CPU_DEVICE_NAME, self.conv_op_standard)
        conv_op_cudnn = _generate_defun_backend(
            api_name, _GPU_DEVICE_NAME, self.conv_op_cudnn)
        res = conv_op_standard(**params) 
        function.register(conv_op_cudnn, **params)
```
I got signature mismatches problem and then I tried to output the FunctionDef infos like (You can find how I output the FunctionDef in function_api_info.cc from the above link.)
```
UUU sig1:
1 self     false
20 inp     false
UUU sig1 output:
1 identity     false
UUU sig2:
1 self     false
20 inp     false
3 required_space_to_batch_paddings_block_shape     false
3 required_space_to_batch_paddings_base_paddings     false
UUU sig2 output:
1 identity     false
```
sig1 is my changed func (conv_op_cudnn) and the inputs only contain the input and filter (The first number is the dtype and I think the shown names are wrongly presented. Since I am using a callable class here, it treats the first name as the input. The correct ones should be `inp` and `filter`.) For the sig2 (conv_op_standard), it has two more inputs `block_shape` and `base_paddings` which are used in the transformation (https://github.com/tensorflow/tensorflow/blob/fd0ba22c25322a709f9a691b18ceffd114857197/tensorflow/python/ops/nn_ops.py#L613). So the signatures of the two functions mismatch and hinder the correct implementation selection(?). Do you have any idea how to deal with this case?

Reply:
I think the expectation of implementation selector is that all the alternatives should have same function signature (eg take same amount of tensor inputs, and return same outputs, wrt to dtypes and shapes). Mathematically, functions should have same behavior, so having different input/output signature seems violating that rule. 

In your case, are required_space_to_batch_paddings_block_shape and required_space_to_batch_paddings_base_paddings tensor inputs? From the name, they seems to be python constants.

The walk around I can think of is feed that two extract tensor to sig1 as well, and let the function body ignore it.

Reply:
Do you mean I can pass these two params to `**params` of conv_op_cudnn? But if conv_op_cudnn doesn't actually use it, wouldn't they be ignored in the FuncDef?

Reply:
If you pass tensors cross the boundary of tf.function, i think they will tracked as signature, regardless whether its being used within the function body.

More importantly, if the cpu function need to take those two tensor param, how could it ensure the cpu function behave the same as the cudnn function?

Reply:
Btw, unused tensor might be optimized by function optimizer in the grappler step, but that happens after the implementation selector.

Reply:
My understanding is that for the CPU function, it needs to prepare more tensor params for the preprocessing step. For example, CPU function accept `padding` (SAME or VALID) as input. Then, it will create a tensor `base_paddings` accordingly. This tensor will be the input param to `array_ops.required_space_to_batch_paddings`. In comparison, cuDNN function directly call the gen_nn_ops.convXd() with the `padding` param.

Reply:
If the CPU function could behave differently than GPU function, eg it can support more settings when config differently, one work around I can think of is give a wrapper function to CPU with the fixed param, so that it will behave same as GPU, and the wrapper function will just call the original CPU function with fixed param:

```python
def original_cpu(inputs, param1, param2):
   ......

def cpu(inputs, param1):
    return orginal_cpu(inputs, param1, param2='padding')

def gpu(inputs, param1):
  .....

# register cpu() and gpu().
```

Reply:
@qlzh727, Yes, I have redesign this PR in the latest commit. PTAL.

Reply:
@reedwm @qlzh727 , I followed the suggestion 1 and redesigned the PR. However, I noticed that many tests are failed. After some digging, I found this might be caused by the incompatibility of the `with tf.device()` scope and the implementation selector. For example, we have a model script like
```
with tf.device(cpu):
  output = op(input)
```
Before this PR, the op will use CPU implementation and every optimization is about CPU.
However, after my changes (calling the CPU op and registering GPU op), I found the optimizations become GPU related. But during the runtime, the invoked kernel is the CPU one. So, for instance, when the op is convolution and the test is NHWC + CPU, we get the error like `NCHW is not implemented by CPU`, indicating the optimizer turns the node to GPU and then call CPU kernel.

I am not sure if you have seen this case before or did I miss something? Thx.

Reply:
Humm, since the GPU function is registered in the graph, the function optimizer will go through it. Having said that, the implementation selector will not swap to use the GPU function since the device placement is CPU, so it will eventually run the CPU function.

Could u turn on the vlog for implemenetation selector? There are debug log when it actually device to do the swap.

Reply:
So, what does the implementation selector actually do? It sounds like the graph has both CPU and GPU op function before and after the implementation selector. Correct?

Reply:
And will the following layout optimizer view this node as CPU or GPU node (because it needs to determine using NHWC or NCHW)?

Reply:
In the main graph, there is a op (partitioned_call or stateful_partitioned_call), which will invoke the function. The implementation_selector will update the op to point to the correct function.

In short, there are several grappler stage:
1. implementation_selector make selection for correct functions to use
2. function_optimizer optimize any function body present, regardless whether it is used or not.
3. model pruner, remove any unused nodes (ops) or function.
....

These steps will go a few rounds, so you might see your GPU function is optimized, but eventually it is deleted since its never used.

Reply:
I think I fixed another `signature mismatches` issue. The GPU test looks good and I can see the selector selects the correct function. But the CPU test still fails and encounters the `NCHW is not implemented on CPU`. 
The vlog from GPU test:
```
2019-10-23 18:04:00.348016: I tensorflow/core/grappler/optimizers/implementation_selector.cc:244] Op convolution/PartitionedCall runs on /job:localhost/replica:0/task:0/device:GPU:0 = (GPU)
2019-10-23 18:04:00.348043: I tensorflow/core/grappler/optimizers/implementation_selector.cc:257] Swapping: __inference_function_20 TO: __inference_function_35
2019-10-23 18:04:00.351491: I tensorflow/core/grappler/optimizers/implementation_selector.cc:141] Node def before swap is: name: "convolution/PartitionedCall"
op: "PartitionedCall"
input: "Const"
input: "Const_1"
device: "/job:localhost/replica:0/task:0/device:GPU:0"
attr {
  key: "Tin"
  value {
    list {
      type: DT_FLOAT
      type: DT_FLOAT
    }
  }
}
attr {
  key: "Tout"
  value {
    list {
      type: DT_FLOAT
    }
  }
}
attr {
  key: "_gradient_op_type"
  value {
    s: "PartitionedCall-21"
  }
}
attr {
  key: "config"
  value {
    s: ""
  }
}
attr {
  key: "config_proto"
  value {
    s: "\n\007\n\003GPU\020\002\n\007\n\003CPU\020\0012\007*\0030,1J\0008\001"
  }
}
attr {
  key: "executor_type"
  value {
    s: ""
  }
}
attr {
  key: "f"
  value {
    func {
      name: "__inference_function_20"
    }
  }
}

2019-10-23 18:04:00.351635: I tensorflow/core/grappler/optimizers/implementation_selector.cc:204] Node def after swap is: name: "convolution/PartitionedCall"
op: "PartitionedCall"
input: "Const"
input: "Const_1"
device: "/job:localhost/replica:0/task:0/device:GPU:0"
attr {
  key: "Tin"
  value {
    list {
      type: DT_FLOAT
      type: DT_FLOAT
    }
  }
}
attr {
  key: "Tout"
  value {
    list {
      type: DT_FLOAT
    }
  }
}
attr {
  key: "_gradient_op_type"
  value {
    s: "PartitionedCall-21"
  }
}
attr {
  key: "config"
  value {
    s: ""
  }
}
attr {
  key: "config_proto"
  value {
    s: "\n\007\n\003GPU\020\002\n\007\n\003CPU\020\0012\007*\0030,1J\0008\001"
  }
}
attr {
  key: "executor_type"
  value {
    s: ""
  }
}
attr {
  key: "f"
  value {
    func {
      name: "__inference_function_35"
    }
  }
}
```

Reply:
The vlog from CPU test:
```
2019-10-23 18:09:08.886718: I tensorflow/core/grappler/optimizers/implementation_selector.cc:244] Op convolution/PartitionedCall runs on /job:localhost/replica:0/task:0/device:CPU:0 = (CPU)
2019-10-23 18:09:08.889179: I tensorflow/core/grappler/optimizers/implementation_selector.cc:286] Skipping optimization since lib_info is empty
2019-10-23 18:09:08.890560: I tensorflow/core/grappler/optimizers/implementation_selector.cc:286] Skipping optimization since lib_info is empty
2019-10-23 18:09:08.891281: I tensorflow/core/grappler/optimizers/implementation_selector.cc:286] Skipping optimization since lib_info is empty
2019-10-23 18:09:08.892175: I tensorflow/core/grappler/optimizers/implementation_selector.cc:286] Skipping optimization since lib_info is empty
2019-10-23 18:09:08.892770: I tensorflow/core/grappler/optimizers/implementation_selector.cc:286] Skipping optimization since lib_info is empty
2019-10-23 18:09:08.893747: I tensorflow/core/grappler/optimizers/implementation_selector.cc:286] Skipping optimization since lib_info is empty
2019-10-23 18:09:08.894369: I tensorflow/core/grappler/optimizers/implementation_selector.cc:286] Skipping optimization since lib_info is empty
2019-10-23 18:09:08.896989: I tensorflow/core/grappler/optimizers/implementation_selector.cc:286] Skipping optimization since lib_info is empty
2019-10-23 18:09:08.897793: I tensorflow/core/grappler/optimizers/implementation_selector.cc:286] Skipping optimization since lib_info is empty
2019-10-23 18:09:08.899079: I tensorflow/core/grappler/optimizers/implementation_selector.cc:286] Skipping optimization since lib_info is empty
2019-10-23 18:09:08.903104: I tensorflow/core/grappler/optimizers/implementation_selector.cc:286] Skipping optimization since lib_info is empty
2019-10-23 18:09:08.907821: I tensorflow/core/grappler/optimizers/implementation_selector.cc:286] Skipping optimization since lib_info is empty
ERROR:tensorflow:[_Derived_] 2 root error(s) found.
  (0) Unimplemented:  The Conv2D op currently only supports the NHWC tensor format on the CPU. The op was given the format: NCHW
	 [[{{node convolution}}]]
  (1) Unimplemented:  The Conv2D op currently only supports the NHWC tensor format on the CPU. The op was given the format: NCHW
	 [[{{node convolution}}]]
	 [[convolution/_4]]
0 successful operations.
0 derived errors ignored.
	 [[convolution/PartitionedCall]]
```

Reply:
If the cpu function only supports NHWC and no GPU is available, when feed input with NCHW, what's the expected behavior? like no function can handle this case?

Reply:
I checked the functions called by the GPU and they are NCHW, although the input is NHWC. So, I think the following layout optimizer must convert the format from NHWC to NCHW for GPUs.

But I am still confused why the CPU test makes the conversion. You mentioned that both CPU/GPU ops will exist in the graph and both CPU/GPU optimizations will be applied on them. Shouldn't it be like this after the optimization (before model prune):
CPU: NHWC-> CPU op -> NHWC
GPU: NHWC -> NCHW -> GPU op -> NCHW -> NHWC
And after the model prune, we pick the CPU one and keep using NHWC if we have CPU. Please let me know if my understanding is correct or not.

Reply:
Yes, I think it will error out when NCHW inputs is given and only CPU is avail (also layout grappler pass is disabled).

Reply:
I am not very sure about the layout optimizer here, eg whether it will try to convert the CPU function body with NHWC -> NCHW -> CPU op (when gpu device is visible, but maybe not placed on GPU). Also this case is bit tricky since it is in a function body, not in main graph. 

Could u check the log on layout optimizer and see if it does any conversion from NHWC -> NCHW for the CPU test case?

Reply:
Yes, I can see that the layout optimizer indeed makes the conversion from NHWC->NCHW for CPU tests.

Reply:
Then the layout optimizer probably need some update, eg it need to check that the op is placed on CPU, rather than just check if GPU is available.

Reply:
Right, the layout optimizer, it uses https://github.com/tensorflow/tensorflow/blob/d0b08ec92d9355f90ab6b2ad7e002f1b7d02e186/tensorflow/core/grappler/optimizers/generic_layout_optimizer.cc#L50 to get the GPU number.

Reply:
It seems it checks the cluster is placed on CPU or GPU. My understanding the cluster is a group of ops. So, this should be sth like `op.GetDevices()`?

Reply:
> 1. implementation_selector make selection for correct functions to use

@qlzh727 On second thought, for the CPU tests, will this step make the Partition_Call choose the the CPU func? If yes, where is the GPU func then and does it get removed? 

I experimented the following code with only CPU func registered. And this still fails for the CPU tests. 
```
        api_name = 'conv_' + str(uuid.uuid4())
        conv_op_standard = _generate_defun_backend(
            api_name, _CPU_DEVICE_NAME, self.conv_op_standard)
        res = conv_op_standard(**params)
```
But, it works only when function isn't used.
```
        res = self.conv_op_standard(**params)
```
So, I think the layout optimizer or other optimizers may not work well with the Partition_Call/func stuff. Can you provide some advice?

Reply:
I further narrowed down the problem. It seems that the function registering mechanism disables the device placement process over each op inside the func.

For example, originally, we have op1->op2->op3, and for CPU test, the device placement can assign the "CPU device" attribute to each of the ops. Then, the layout optimization pass is able to use the `node.device()` to determine the correct layout format.
https://github.com/tensorflow/tensorflow/blob/8845f2c52c1c7ba8db63a64493630a50f4f4a303/tensorflow/core/grappler/optimizers/generic_layout_optimizer_transposer.cc#L1641 

However, when we wrap up op1->op2->op3 into func1 (using _generate_defun_backend), the device placement step is somehow interrupted and causes the `node.device()` empty for each of the ops. Thus, the layout optimization pass has to use `virtual_placer->get_canonical_device_name(node)` to get the device info, which in turn uses the default device (GPU:0). 
https://github.com/tensorflow/tensorflow/blob/8845f2c52c1c7ba8db63a64493630a50f4f4a303/tensorflow/core/grappler/optimizers/generic_layout_optimizer_transposer.cc#L1640
https://github.com/tensorflow/tensorflow/blob/8845f2c52c1c7ba8db63a64493630a50f4f4a303/tensorflow/core/grappler/costs/virtual_placer.cc#L80

So, one potential solution is that during the implementation selector pass, we could also assign each op with the device info based on _DEFUN_DEVICE_ATTRIBUTE.